File: priv/combined_metrics/samples/code_smells/no_fixme_comments/good/importer.ex
defmodule Importer do
  @moduledoc "Handles importing data from CSV and external sources"

  def import_csv(path) do
    case File.read(path) do
      {:ok, ""} ->
        {:error, :empty_file}

      {:ok, content} ->
        rows =
          content
          |> String.split("\n", trim: true)
          |> Enum.map(&parse_row/1)
          |> Enum.reject(&is_nil/1)

        {:ok, rows}

      {:error, reason} ->
        {:error, reason}
    end
  end

  def parse_row(line) do
    case String.split(line, ",") do
      [id, name, email] ->
        %{id: id, name: name, email: email}

      _ ->
        nil
    end
  end

  def import_users(rows) do
    rows
    |> Enum.map(&insert_user/1)
    |> Enum.split_with(&match?({:ok, _}, &1))
    |> then(fn {ok, errors} -> {:ok, length(ok), length(errors)} end)
  end

  def validate_row(%{email: email} = row) do
    normalized = String.downcase(email)

    if String.match?(normalized, ~r/^[^\s@]+@[^\s@]+\.[^\s@]+$/) do
      {:ok, %{row | email: normalized}}
    else
      {:error, :invalid_email}
    end
  end

  def deduplicate(rows) do
    rows
    |> Enum.map(fn row -> %{row | email: String.downcase(row.email)} end)
    |> Enum.group_by(& &1.email)
    |> Enum.map(fn {_email, [first | _rest]} -> first end)
  end

  def import_from_api(source_url) do
    with {:ok, data} <- fetch_remote(source_url),
         {:ok, parsed} <- parse_api_response(data) do
      {:ok, parsed}
    end
  end

  def transform_row(row, field_map) do
    Enum.reduce(field_map, %{}, fn {src, dst}, acc ->
      Map.put(acc, dst, Map.get(row, src))
    end)
  end

  def write_results(results, output_path) do
    backup_path = output_path <> ".bak"

    with :ok <- maybe_backup(output_path, backup_path),
         content = Enum.map_join(results, "\n", &format_result/1),
         :ok <- File.write(output_path, content) do
      :ok
    end
  end

  defp maybe_backup(path, backup) do
    if File.exists?(path), do: File.copy(path, backup), else: :ok
  end

  defp insert_user(row), do: {:ok, row}
  defp fetch_remote(_url), do: {:ok, []}
  defp parse_api_response(data), do: {:ok, data}
  defp format_result(result), do: inspect(result)
end

File: priv/combined_metrics/samples/code_smells/no_fixme_comments/good/importer.rb
# Handles importing data from CSV and external sources.

module Importer
  EMAIL_REGEX = /\A[^\s@]+@[^\s@]+\.[^\s@]+\z/.freeze

  module_function

  def import_csv(path)
    content = File.read(path)
    return { error: :empty_file } if content.empty?

    rows = content
           .split("\n")
           .reject { |line| line.strip.empty? }
           .map { |line| parse_row(line) }
           .compact

    { ok: rows }
  end

  def parse_row(line)
    parts = line.split(",")
    return nil unless parts.length == 3

    id, name, email = parts
    { id: id, name: name, email: email }
  end

  def import_users(rows)
    results = rows.map { |row| insert_user(row) }
    ok, errors = results.partition { |r| r.key?(:ok) }
    { ok: ok.length, errors: errors.length }
  end

  def validate_row(row)
    normalized = row[:email].downcase
    if EMAIL_REGEX.match?(normalized)
      { ok: row.merge(email: normalized) }
    else
      { error: :invalid_email }
    end
  end

  def deduplicate(rows)
    rows
      .map { |row| row.merge(email: row[:email].downcase) }
      .group_by { |row| row[:email] }
      .map { |_email, group| group.first }
  end

  def import_from_api(source_url)
    fetched = fetch_remote(source_url)
    return fetched if fetched.key?(:error)

    parse_api_response(fetched[:ok])
  end

  def transform_row(row, field_map)
    field_map.each_with_object({}) do |(src, dst), acc|
      acc[dst] = row[src]
    end
  end

  def write_results(results, output_path)
    backup_path = "#{output_path}.bak"
    FileUtils.cp(output_path, backup_path) if File.exist?(output_path)

    content = results.map { |r| format_result(r) }.join("\n")
    File.write(output_path, content)
    { ok: true }
  end

  def insert_user(row)
    { ok: row }
  end

  def fetch_remote(_url)
    { ok: [] }
  end

  def parse_api_response(data)
    { ok: data }
  end

  def format_result(result)
    result.inspect
  end
end
